tests/unit/test_cycle.cpp

```diff
@@ -0,0 +1,179 @@
+                                                     values.cend());
+
+        for (auto i = 0u; i < values.size(); ++i) {
+            EXPECT_EQ(values[i], cycle_iter[values.size() + i]);
+        }
+    }
+}
+
+TEST(cycle_iterator, decrement) {
+    std::vector<int> values = { 4, 2, 3 };
+
+    {
+        // test operator--
+        auto cycle_iter = util::make_cyclic_iterator(values.cbegin(),
+                                                     values.cend());
+        auto cycle_iter_copy = cycle_iter;
+
+        auto values_size = values.size();
+        for (auto i = 0u; i < 2*values_size; ++i) {
+            --cycle_iter;
+            cycle_iter_copy--;
+            auto val = values[values_size - i%values_size - 1];
+            EXPECT_EQ(val, *cycle_iter);
+            EXPECT_EQ(val, *cycle_iter_copy);
+        }
+    }
+
+    {
+        // test operator[]
+        auto cycle_iter = util::make_cyclic_iterator(values.cbegin(),
+                                                     values.cend());
+        auto values_size = values.size();
+        for (auto i = 0; i < 2*values_size; ++i) {
+            auto pos = i % values_size;
+            pos = pos ? values_size - pos : 0;
+            EXPECT_EQ(values[pos], cycle_iter[-i]);
+        }
+    }
+}
+
+TEST(cycle_iterator, carray) {
+    int values[] = { 4, 2, 3 };
+    {
+        // test operator++
+        auto cycle_iter = util::make_cyclic_iterator(util::cbegin(values),
+                                                     util::cend(values));
+        auto cycle_iter_copy = cycle_iter;
+
+        auto values_size = util::size(values);
+        for (auto i = 0u; i < 2*values_size; ++i) {
+            EXPECT_EQ(values[i % values_size], *cycle_iter);
+            EXPECT_EQ(values[i % values_size], *cycle_iter_copy++);
+            ++cycle_iter;
+        }
+    }
+}
+
+TEST(cycle_iterator, sentinel) {
+    using testing::null_terminated;
+
+    auto msg = "hello";
+    auto cycle_iter = util::make_cyclic_iterator(msg, null_terminated);
+
+    auto msg_len = std::string(msg).size();
+    for (auto i = 0u; i < 2*msg_len; ++i) {
+        EXPECT_EQ(msg[i % msg_len], *cycle_iter++);
+    }
+}
+
+TEST(cycle, cyclic_view) {
+    std::vector<int> values = { 4, 2, 3 };
+
+    {
+        std::vector<int> values_new(10, -1);
+
+        auto cycle = util::cyclic_view(values);
+        auto values_new_iter = values_new.begin();
+        for (auto const& v : cycle) {
+            if (values_new_iter == values_new.end()) {
+                break;
+            }
+
+            *values_new_iter = v;
+            ++values_new_iter;
+        }
+
+        auto i = 0;
+        for (auto const& v : values_new) {
+            EXPECT_EQ(values[i % values.size()], v);
+            ++i;
+        }
+    }
+
+    {
+        std::vector<int> values_new(10, -1);
+
+        auto cycle_iter = util::cyclic_view(values).cbegin();
+        for (auto &v : values_new) {
+            v = *cycle_iter++;
+        }
```

Comment:
I was trying now to rewrite this in a more C++ fashion as follows:
```c++
        std::vector<int> values_new;
        std::copy_n(util::cyclic_view(values).cbegin(), 10,
                    std::back_inserter(values_new));

        EXPECT_EQ(10, values_new.size());
```

Interestingly, the above expectation fails. More interestingly, I realised that Clang 3.9, which I'm using, implements `copy_n()` by first getting an "end" iterator as `std::next(begin, 10)` and then trying to copy elements. However, moving the cycle iterator 10 times, will bring you just to the second element, that's why you end up with only one element in the new vector. The proposed implementation [here](http://en.cppreference.com/w/cpp/algorithm/copy_n) works fine with the above code and meets in any case the expected criterion of `copy_n()` of "Copies exactly count values from the range beginning at first to the range beginning at result". I can understand Clang wouldn't expect a "cyclic" range to be passed through `copy_n()` but I don't believe my view is at fault here. Opinions? 

I will try it also with another compiler. Alternatively, if we want to `copy_n()` portably (?), I should augment the internal state of the `cyclic_iterator` by a `num_wrap_arounds` to keep track of the "wrap-arounds". Gets ugly?